packages/medusa-payment-paypal/src/core/paypal-http-client.ts

```diff
@@ -0,0 +1,132 @@
+import axios, { AxiosInstance, AxiosRequestConfig, Method } from "axios"
+import { PaypalApiPath, PaypalSdkOptions } from "./types"
+import { Logger } from "@medusajs/medusa"
+
+export class PaypalHttpClient {
+  protected readonly baseUrl_: string = "https://api-m.paypal.com/v2"
+  protected readonly httpClient_: AxiosInstance
+  protected readonly options_: PaypalSdkOptions
+  protected readonly logger_?: Logger
+  protected accessToken_: string
+
+  constructor(options: PaypalSdkOptions) {
+    this.options_ = options
+
+    this.logger_ = options.logger
+
+    if (options.useSandbox) {
+      this.baseUrl_ = "https://api-m.sandbox.paypal.com"
+    }
+
+    const axiosInstance = axios.create({
+      baseURL: this.baseUrl_,
+    })
+
+    const proxifiyMethodTarget = ["request", "post", "get"]
+
+    this.httpClient_ = new Proxy(axiosInstance, {
+      // Handle automatic retry mechanism
+      get: (target, prop) => {
+        // @ts-ignore
+        // eslint-disable-next-line prefer-rest-params
+        const original = Reflect.get(...arguments)
+
+        if (proxifiyMethodTarget.includes(prop as string)) {
+          return this.retryIfNecessary(original)
+        }
+
+        return original
+      },
+    })
+  }
+
+  /**
+   * Run a request and return the result
+   * @param url
+   * @param data
+   * @param method
+   * @protected
+   */
+  async request<T, TResponse>({
+    url,
+    data,
+    method,
+  }: {
+    url: string
+    data?: T
+    method?: Method
+  }): Promise<TResponse> {
+    return await this.httpClient_
+      .request({
+        method: "POST",
+        url,
+        headers: {
+          "Content-Type": "application/json",
+          Authorization: `Bearer ${this.accessToken_}`,
+        },
+        data,
+      })
+      .then((res) => res.data)
```

Comment:
question: Should we just return the response object directly? If not, would also recommend returning the "body" of the error response as well.

Reply:
I ll manage that tomorrow, at the moment it is not totally finished for the core part 🫣but thanks for all the feedback it will be integrated for sure. Also, at the moment the retry catch but i do not re throw (testing purpose) but it will re trhow for the processor to handle the error on its own with the actual error strategy